### ISSUE TITLE
Tagging scripts as free code for LibreJS compatibility

### ISSUE THREAD

**Issue by strypey** [loomio/loomio#2331]
Kia ora te whānau o Loomio

My first ever Issue created on GITHub!

Can coders please tag nontrrivial Loomio scripts so they are recognised as the free code they are (I presume) by LibreJS:
https://www.gnu.org/software/librejs/

This basically involves a web link to the full source code of the script and reference to the AGPL or whichever license covers them? More details on freeing your scripts here:
https://www.gnu.org/software/librejs/free-your-javascript.html
        
More background on the problems created by nonfree JS that this is intended to solve:
https://www.gnu.org/philosophy/javascript-trap.html
        
I'm highlighting this Issue in support of the FSF 'Free Javascript Campaign':
https://www.fsf.org/campaigns/freejs
        
I thought I'd pluck the low-hanging fruit, and start with you lovely folks :)
        
He mihi mahana
Strypey

**Comment by robguthrie** [comment]
Hi @strypey,

As you know, all the software used in Loomio is free and the license is easy to find in our repo.

Personally I find those pages you've linked and the whole javascript-trap campaign very hard to understand and I've given up trying to read them for now.

I started by thinking .. ok I'll give this a go... how do I embed a license link in our app. And after 10 minutes I just felt confused and gave up.

Can you explain what I need to do?

**Comment by strypey** [comment]
Kia ora @robguthrie,

"Can you explain what I need to do?"

Sadly I can't, the implementation details are beyond my technical knowledge. Thanks for giving it a try. I have sent an email to the LibreJS mailing list linking to this page, and asking if any of them are able to help:
https://lists.gnu.org/mailman/listinfo/bug-librejs

The link is the one which explains what to do. If you could clarify where it's losing you, I'll do my best to help, or find someone who can:
https://www.gnu.org/software/librejs/free-your-javascript.html

I'm aware that JavaScript Trap article is long and detailed, and I know you folk are busy, so I'll try a TL;DR:
JavaScript in web pages allows the browser to run programs on their computer, without their knowledge or permission, many of which are proprietary. Users who value software freedom often turn JavaScript off to avoid this (and improve browser security), but it means they can't properly use free code server software like Loomio. Scripts which are licensed as free code (like those in Loomio) can be modified to identify to the browser, in a standardized way, which license covers their code, and where copies of that license and full source code can be found. Then, the user can be empowered to use the sites serving free scripts with confidence, and make more granular decisions about if and when to allow proprietary scripts to run on their machine.

You may feel that this whole issue is nit-picking and not worth your time, and if so, I totally respect your right to hold that opinion. However, if you agree that this is problem worth trying to solve, LibreJS is being developed as a tool to facilitate that. Keep in mind that it is new and still experimental. The sysadmins at Quitter.se got their scripts working with LibreJS in early June, then something changed, and now they're not, so I'd suggest looking at it as a medium-to-long term priority.

**Comment by rdbartlett** [comment]
Thanks for the TLDR Strypey, helps a lot.

I think it makes sense to leave this issue open so someone can pick up the
task if they have the space and inclination :)

**Comment by robguthrie** [comment]
I'm going to close this, as I want to keep the issues list relevant to stuff we want to be thinking about.

If someone wants to do this work they are welcome to open a pull request.

Thanks for letting us know about this project @strypey

**Comment by strypey** [comment]
De nada. Here's another useful essay about the software freedom issues JavaScript creates (about 2 pages and fairly simple to follow):
https://onpon4.github.io/other/kill-js/

TL;DR
Javascript "programs", even if they are under a free code license, are not free code software in the sense that the user can't modify them, and use their modified version in place of the default one supplied by the site. Now, in practice, if the JS code is maintained by an open source project like Loomio, a user can do this in a sense by sending you guys a pull request. But this assumes you accept it. If you don't, there's no way for users to fork a script, a use their own fork instead of the ones you guys supply. Unless they set up their own server, but unless each user operates their own server (which would be pointless unless they could federate - which would be cool), each user is forced to run the version of each script that the server admin chooses. 

Now, none of this is Loomio's fault :) It's just the way web technologies have evolved, and I'm guessing a lot of the scripts a Loomio site serves aren't even built by Loomio, but are part of various dependencies you make use of. However if creative ways are suggested to make it possible for users to have more control over the JS "programs" that web apps like Loomio rely on running in web browsers, I'm sure you folk would be keen to support them.

**Comment by robguthrie** [comment]
I'm not going to talk about Librejs anymore. All our scripts are libre.

**Comment by strypey** [comment]
Fair enough @robguthrie . Just sharing relevant info for future reference.  I hope it didn't seem like I was demanding that you (or anyone else) respond.

**Comment by robguthrie** [comment]
Sorry to be blunt like that - but I think LibreJS offers very little value to users of open source apps like ours, and users of non-free apps don't care anyway. I'm terrified to share my views because then I'll be involved in discussing it further, which is what I want to avoid, because it consumes my time to write more good libre code.

**Comment by strypey** [comment]
No more walls of text about "values" and "principles", I promise! Peace? :)

**Comment by gdpelican** [comment]
@strypey you rock and keep us honest and that's a great thing. I just prefer that discussion about principles happen on Loomio, and discussion about code happen in Github. <3

**Comment by strypey** [comment]
Understood. I opened this issue originally after being asked to by @rdbartlett . I added the info Robert was responding to in case it is helpful if anyone (including myself) decides LibreJs compatibility is something they want to work on. Like I said no more walls of text about "values" and "principles", I promise! Especially in GITHub.

**Comment by robguthrie** [comment]
oh man.. I'm not mad or miffed at you @strypey. Thanks for being around. I just can't stand talking about librejs :)

**Comment by strypey** [comment]
Sweetbix :)
